### PR TITLE
use systemd to notify shutdown for watch commands

### PR DESCRIPTION
**Pull request by domenkozar**
Allows clean shutdown when used with systemd. cc @mic92 for testing

### CODE REVIEW

File: cachix/src/Cachix/Client/PushQueue.hs [hunk 1]
```diff
@@ -95,6 +96,7 @@ queryLoop workerState pushqueue pushParams = do
 exitOnceQueueIsEmpty :: IO () -> Async () -> Async () -> QueryWorkerState -> PushWorkerState -> IO ()
 exitOnceQueueIsEmpty stopProducerCallback pushWorker queryWorker queryWorkerState pushWorkerState = do
   putText "Stopped watching /nix/store and waiting for queue to empty ..."
+  Systemd.notifyStopping
```
Comment (Mic92) [inline cachix/src/Cachix/Client/PushQueue.hs:99]: Is this the haskell counterpart for?

```
           If a service of Type=notify sends "EXTEND_TIMEOUT_USEC=...", this
           may cause the stop time to be extended beyond TimeoutStopSec=. The
           first receipt of this message must occur before TimeoutStopSec= is
           exceeded, and once the stop time has extended beyond
           TimeoutStopSec=, the service manager will allow the service to
           continue to stop, provided the service repeats
           "EXTEND_TIMEOUT_USEC=..."  within the interval specified, or
           terminates itself (see sd_notify(3)).
```

This is from `systemd.service(5)` manpage.

Comment (Mic92) [reply cachix/src/Cachix/Client/PushQueue.hs:99]: Seems to be not implemented yet but https://hackage.haskell.org/package/systemd-2.3.0/docs/src/System.Systemd.Daemon.html#notifyStatus
allows to implement that. My haskell foo is not strong enough to come up with more concrete suggestions.

Comment (Mic92) [reply cachix/src/Cachix/Client/PushQueue.hs:99]: But reading further below looks like this function would be the right place to extend the timeout.

**Comment by domenkozar** [comment]
@Mic92 PTAL :)

File: cachix/src/Cachix/Client/PushQueue.hs [hunk 2]
```diff
@@ -112,6 +116,8 @@ exitOnceQueueIsEmpty stopProducerCallback pushWorker queryWorker queryWorkerStat
           cancelWith queryWorker StopWorker
           cancelWith pushWorker StopWorker
         else do
+          -- extend shutdown for another 90s
+          Systemd.notify False $ "EXTEND_TIMEOUT_USEC=" <> show (90 * 1000 * 1000)
```
Comment (Mic92) [inline cachix/src/Cachix/Client/PushQueue.hs:120]: I was not sure how systemd would interpret this timeout but what it does is according to the source:
`current_time + EXTEND_TIMEOUT_USEC`. So this implementation seems correct: if there is no cachix update for 90s
than the service is killed.

Comment (domenkozar) [reply cachix/src/Cachix/Client/PushQueue.hs:120]: That was my interpretation from `the service manager will allow the service to
           continue to stop, provided the service repeats
           "EXTEND_TIMEOUT_USEC=..."  within the interval specified, or
           terminates itself (see sd_notify(3)).`

Comment (domenkozar) [reply cachix/src/Cachix/Client/PushQueue.hs:120]: @Mic92 if this looks good to you I'll merge it and you can give it a try :)

Comment (Mic92) [reply cachix/src/Cachix/Client/PushQueue.hs:120]: It does.

**Pull request merged**